Greenplum: PXF CLI (#243)
pivotal-legacy/blog: content/post/gpdb-pxf-cli.md

Using CATALINA_PID:    /usr/local/greenplum-db-devel/pxf/run/catalina.pid
Tomcat stopped.
~~~

Checking the cluster's status from master again,

~~~bash
[gpadmin@mdw ~]$ pxf cluster status
~~~

~~~terminal
Checking status of PXF servers on 2 hosts...
ERROR: PXF is not running on 1 out of 2 hosts
sdw1 ==> Checking if tomcat is up and running...
ERROR: PXF is down - tomcat is not running...
~~~

The output displays the truncated `stderr` message for whichever host had a non-zero return code. This is a pattern that we follow for all cluster-level commands.

## Workflow 3: Syncing configurations

To confirm that PXF is working properly on the cluster, let's [create a table](https://gpdb.docs.pivotal.io/5210/pxf/objstore_text.html) that points to [Amazon S3](https://aws.amazon.com/s3/), then query that table.

First generate a file with some random words to put in a file on S3.

~~~bash
[gpadmin@mdw ~]$ shuf -n7 /usr/share/dict/words | awk '{print NR","$1}' > random-words.csv
[gpadmin@mdw ~]$ cat random-words.csv
~~~

~~~terminal
1,savvying
2,teleophore
3,automaker
4,vitriolling
5,nonadministrable
6,switch-over
7,stegnosisstegnotic
~~~

> Now we have some data to query using PXF. Upload it to S3 using the [AWS Command Line Interface](https://aws.amazon.com/cli/):

Reviewer: ```suggestion
Now we have successfully generated data to query using PXF. You can store the file in S3 using the [AWS Command Line Interface](https://aws.amazon.com/cli/):
```

~~~SQL
[gpadmin@mdw ~]$ psql template1
template1=# CREATE EXTENSION pxf; -- register the PXF extension, needed once per database
template1=# CREATE EXTERNAL TABLE random_words (id int, word text)
template1=#   LOCATION ('pxf://bucket-name/path/random-words.csv?PROFILE=s3:text&SERVER=my_s3')
template1=#   FORMAT 'csv';
~~~

~~~terminal
CREATE EXTERNAL TABLE
~~~

The location clause above specifies a server named `my_s3`, but we have not yet configured this server. When we query our table, PXF returns an exception:

~~~SQL
template1=# SELECT * FROM random_words;
~~~

~~~terminal
ERROR:  remote component error (500) from '127.0.0.1:5888':  type  Exception report   message   javax.servlet.ServletException: org.apache.hadoop.fs.s3a.AWSClientIOException: doesBucketExist on bucket-name: com.amazonaws.AmazonClientException: No AWS Credentials provided by BasicAWSCredentialsProvider EnvironmentVariableCredentialsProvider InstanceProfileCredentialsProvider
~~~

> This error indicates that we have no credentials for S3. This makes sense, since we have not configured a server in `$PXF_CONF/servers/my_s3`, and we didn't provide credentials as parameters in the location clause when we created the table.

Reviewer: ```suggestion
This error indicates that PXF is unable to provide credentials to S3. This behavior is  expected, since we have not configured a server in `$PXF_CONF/servers/my_s3`, and we didn't provide credentials as parameters in the location clause when we created the table.
```

> Create a server configuration file at `$PXF_CONF/servers/my_s3/s3-site.xml` and edit the file to include your S3 access key and secret (see [here](https://gpdb.docs.pivotal.io/5210/pxf/objstore_cfg.html) for details).

Reviewer: ```suggestion
Create a server configuration file at `$PXF_CONF/servers/my_s3/s3-site.xml` and edit the file to include your S3 access and secret keys (see [here](https://gpdb.docs.pivotal.io/5210/pxf/objstore_cfg.html) for details).
```

~~~bash
[gpadmin@mdw ~]$ mkdir "${PXF_CONF}/servers/my_s3"
[gpadmin@mdw ~]$ cp "${PXF_CONF}/templates/s3-site.xml" "${PXF_CONF}/servers/my_s3"
[gpadmin@mdw ~]$ vi "${PXF_CONF}/servers/my_s3/s3-site.xml"
~~~

> During the query, each PXF server reads configuration data from disk on their respective hosts. For this reason, we must get the configuration file onto the segment hosts, not just master. The master host acts as a central repository for this data. In case the master host goes down, the standby master must also have this data.

Reviewer: ```suggestion
During the query, each PXF server reads configuration data from disk on their respective hosts. For this reason, we must get the configuration file onto all segment hosts, not just master. The master host acts as a central repository for this data. In case the master host goes down, the standby master must also have this data.
```

The PXF CLI makes it easy to sync configurations from the master host to all other hosts, including standby master:

~~~bash
[gpadmin@mdw ~]$ pxf cluster sync
~~~

~~~terminal
Syncing PXF configuration files to 3 hosts...
PXF configs synced successfully on 3 out of 3 hosts
~~~

The `cluster sync` operation acts on the `$PXF_CONF/{servers,conf,lib}` directories, making sure that what is on master will be reflected on the master standby and segment hosts.

Let's make sure that the configuration file is now synced onto the remote hosts.

~~~bash
[gpadmin@mdw ~]$ for server in smdw sdw{1,2}; do
> echo -n "${server}: "
> ssh $server ls "$PXF_CONF/servers/my_s3/s3-site.xml"
> done
~~~

~~~terminal
smdw: /usr/share/pxf/servers/my_s3/s3-site.xml
sdw1: /usr/share/pxf/servers/my_s3/s3-site.xml
sdw2: /usr/share/pxf/servers/my_s3/s3-site.xml
~~~

> And if we try our query again, we see that it returns the expected results.

Reviewer: ```suggestion
Let's run our query again. This time the query succeeds and returns the expected results.
```

~~~SQL
template1=# SELECT * FROM random_words;
~~~

~~~terminal
 id |        word
----+--------------------
  1 | savvying
  2 | teleophore
  3 | automaker
  4 | vitriolling
  5 | nonadministrable
  6 | switch-over
  7 | stegnosisstegnotic
(7 rows)
~~~

## Wrapping up

> The PXF CLI utility, and especially the cluster subcommand make it easy to administer PXF instances, even with large numbers of hosts in your cluster.

Reviewer: ```suggestion
The PXF CLI utility, and especially the cluster subcommand make it easy to administer the PXF cluster, even with large numbers of hosts in your cluster.
```
